File: typescript/src/providers/defaults.ts
// SPDX-License-Identifier: Apache-2.0
/**
 * DefaultProvider — extracts schema defaults as a plain object.
 */

import { z } from "zod";
import { getSchemaShape } from "../schema.js";

/**
 * Extracts default values from a zod schema by parsing an empty object.
 *
 * The result is a plain object with all defaults filled in,
 * suitable for use as the base layer in the merge pipeline.
 */
export class DefaultProvider {
  load(schema: z.ZodTypeAny): Record<string, unknown> {
    // Parse an empty object to get all defaults.
    // For AppConfigSchema with .default({}) on sub-objects, parsing {}
    // fills in all defaults recursively.
    const result = schema.parse({});
    return result as Record<string, unknown>;
  }
}

File: python/src/seventhplace/providers/defaults.py
# SPDX-License-Identifier: Apache-2.0
"""DefaultProvider — extracts schema defaults as a dict."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, SecretStr


def _reveal_secrets(data: Any) -> Any:
    """Recursively convert SecretStr instances to their plain string values."""
    if isinstance(data, dict):
        return {k: _reveal_secrets(v) for k, v in data.items()}
    if isinstance(data, SecretStr):
        return data.get_secret_value()
    if isinstance(data, list):
        return [_reveal_secrets(item) for item in data]
    return data


class DefaultProvider:
    """Extracts default values from a Pydantic schema class."""

    def load(self, schema: type[BaseModel]) -> dict[str, Any]:
        """Instantiate schema with no arguments, dump to dict.

        SecretStr values are revealed (converted to plain strings)
        because the merge pipeline works in dict-space with plain values.
        Pydantic re-wraps them into SecretStr at the final validation step.
        """
        instance = schema()
        data = instance.model_dump()
        return _reveal_secrets(data)
